src/Workspaces/Core/Portable/Workspace/WorkspaceChangeEventArgs.cs

```diff
@@ -21,21 +23,47 @@ public class WorkspaceChangeEventArgs : EventArgs
-        /// <see cref="OldSolution"/> and <see cref="NewSolution"/>.
+        /// If linked documents are being changed, there may be multiple events with the same <see cref="OldSolution"/>
+        /// and <see cref="NewSolution"/>.  Will never be <see langword="null"/>, even for a <see
+        /// cref="WorkspaceChangeKind.SolutionAdded"/> notification.
         /// </remarks>
         public Solution OldSolution { get; }
 
         /// <remarks>
-        /// If linked documents are being changed, there may be multiple events with the same
-        /// <see cref="OldSolution"/> and <see cref="NewSolution"/>.
+        /// If linked documents are being changed, there may be multiple events with the same <see cref="OldSolution"/>
+        /// and <see cref="NewSolution"/>. Will never be <see langword="null"/>, even for a <see
+        /// cref="WorkspaceChangeKind.SolutionRemoved"/> notification.
         /// </remarks>
         public Solution NewSolution { get; }
 
-        public ProjectId ProjectId { get; }
-        public DocumentId DocumentId { get; }
+        /// <summary>
+        /// The id of the affected <see cref="Project"/>.  Can be <see langword="null"/> if this is an change unrelated
+        /// to a project (for example <see cref="WorkspaceChangeKind.SolutionReloaded"/>.  Should be non-<see
+        /// langword="null"/> for:
+        /// <list type="bullet">
+        /// <item><see cref="WorkspaceChangeKind.ProjectAdded"/></item>
+        /// <item><see cref="WorkspaceChangeKind.ProjectChanged"/></item>
+        /// <item><see cref="WorkspaceChangeKind.ProjectReloaded"/></item>
+        /// <item><see cref="WorkspaceChangeKind.ProjectRemoved"/></item>
+        /// </list>
+        /// </summary>
+        public ProjectId? ProjectId { get; }
+
+        /// <summary>
+        /// The id of the affected <see cref="Document"/>.  Can be <see langword="null"/> if this is an change unrelated
+        /// to a document (for example <see cref="WorkspaceChangeKind.ProjectAdded"/>. Should be non-<see
+        /// langword="null"/> for:
+        /// <list type="bullet">
+        /// <item><see cref="WorkspaceChangeKind.DocumentAdded"/></item>
+        /// <item><see cref="WorkspaceChangeKind.DocumentChanged"/></item>
+        /// <item><see cref="WorkspaceChangeKind.DocumentInfoChanged"/></item>
+        /// <item><see cref="WorkspaceChangeKind.DocumentReloaded"/></item>
+        /// <item><see cref="WorkspaceChangeKind.DocumentRemoved"/></item>
+        /// </list>
+        /// </summary>
+        public DocumentId? DocumentId { get; }
 
-        public WorkspaceChangeEventArgs(WorkspaceChangeKind kind, Solution oldSolution, Solution newSolution, ProjectId projectId = null, DocumentId documentId = null)
+        public WorkspaceChangeEventArgs(WorkspaceChangeKind kind, Solution oldSolution, Solution newSolution, ProjectId? projectId = null, DocumentId? documentId = null)
         {
             this.Kind = kind;
```

Comment:
>this.Kind = kind; [](start = 12, length = 17)

Since you're changing this and it is a public API ... it should throw ArgumentNullReference if oldSolution or newSolution are null and ArgumentOutOfRange if kind is invalid.

Reply:
ok